File: app/src/main/java/co/tpcreative/supersafe/common/controller/PremiumManager.java
package co.tpcreative.supersafe.common.controller;
import com.anjlab.android.iab.v3.BillingProcessor;
import com.anjlab.android.iab.v3.PurchaseData;
import com.anjlab.android.iab.v3.TransactionDetails;
import com.google.gson.Gson;
import co.tpcreative.supersafe.R;
import co.tpcreative.supersafe.common.services.SuperSafeApplication;
import co.tpcreative.supersafe.common.util.Utils;
import co.tpcreative.supersafe.model.CheckoutItems;
import co.tpcreative.supersafe.model.EnumPurchase;
import co.tpcreative.supersafe.model.User;

public class PremiumManager implements BillingProcessor.IBillingHandler{

    private static PremiumManager instance;
    private String TAG = PremiumManager.class.getSimpleName();

    /*New version*/
    BillingProcessor bp;

    public static PremiumManager getInstance() {
        if (instance == null) {
            instance = new PremiumManager();
        }
        return instance;
    }


    public void onStartInAppPurchase() {
        bp = new BillingProcessor(SuperSafeApplication.getInstance(),Utils.GOOGLE_CONSOLE_KEY, this);
        bp.initialize();
    }

    public void onStop() {
        if (bp != null) {
            bp.release();
        }
    }

    /*New version*/
    @Override
    public void onProductPurchased(String productId, TransactionDetails details) {

    }

    @Override
    public void onPurchaseHistoryRestored() {

    }

    @Override
    public void onBillingError(int errorCode, Throwable error) {

    }

    @Override
    public void onBillingInitialized() {
        /*Checking life time in-app*/
        if (bp.isPurchased(SuperSafeApplication.getInstance().getString(R.string.life_time))){
            final TransactionDetails details = bp.getPurchaseTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.life_time));
            if (details!=null){
                final PurchaseData mPurchaseData = details.purchaseInfo.purchaseData;
                if (mPurchaseData!=null){
                    onCheckout(mPurchaseData,EnumPurchase.LIFETIME);
                }
            }
        }

        /*Checking subscription for period time 6 months*/
        else if (bp.isSubscribed(SuperSafeApplication.getInstance().getString(R.string.six_months))){
            final TransactionDetails details = bp.getSubscriptionTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.six_months));
            if (details!=null){
                final PurchaseData mPurchaseData = details.purchaseInfo.purchaseData;
                if (mPurchaseData!=null){
                    onCheckout(mPurchaseData,EnumPurchase.SIX_MONTHS);
                }
            }
        }

        /*Checking subscription for period time 1 year*/
        else if (bp.isSubscribed(SuperSafeApplication.getInstance().getString(R.string.one_years))){
            final TransactionDetails details = bp.getSubscriptionTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.one_years));
            if (details!=null){
                final PurchaseData mPurchaseData = details.purchaseInfo.purchaseData;
                if (mPurchaseData!=null){
                    onCheckout(mPurchaseData,EnumPurchase.ONE_YEAR);
                }
            }
        }else {
            onCheckout(null,EnumPurchase.NONE);
        }
    }

    private void onCheckout(PurchaseData data, EnumPurchase purchase){
        CheckoutItems mCheckout = Utils.getCheckoutItems();
        Utils.Log(TAG,"Call checkout....");
        Utils.Log(TAG,new Gson().toJson(data));
        switch (purchase){
            case LIFETIME:
                if (mCheckout != null) {
                    mCheckout.isPurchasedLifeTime = Utils.isRealCheckedOut(data.orderId);
                }
                else{
                    mCheckout = new CheckoutItems();
                    mCheckout.isPurchasedLifeTime = Utils.isRealCheckedOut(data.orderId);
                }
                Utils.setCheckoutItems(mCheckout);
                break;
            case SIX_MONTHS:
                if (mCheckout != null) {
                    if (Utils.isRealCheckedOut(data.orderId)){
                        if (data.autoRenewing){
                            mCheckout.isPurchasedSixMonths = true;
                        }else{
                            mCheckout.isPurchasedSixMonths = false;
                        }
                    }else{
                        mCheckout.isPurchasedSixMonths = false;
                    }
                }
                else{
                    mCheckout = new CheckoutItems();
                    if (Utils.isRealCheckedOut(data.orderId)){
                        if (data.autoRenewing){
                            mCheckout.isPurchasedSixMonths = true;
                        }else{
                            mCheckout.isPurchasedSixMonths = false;
                        }
                    }else{
                        mCheckout.isPurchasedSixMonths = false;
                    }
                }
                Utils.setCheckoutItems(mCheckout);
            case ONE_YEAR:
                if (mCheckout != null) {
                    if (Utils.isRealCheckedOut(data.orderId)){
                        if (data.autoRenewing){
                            mCheckout.isPurchasedOneYears = true;
                        }else{
                            mCheckout.isPurchasedOneYears = false;
                        }
                    }else{
                        mCheckout.isPurchasedOneYears = false;
                    }
                }
                else{
                    mCheckout = new CheckoutItems();
                    if (Utils.isRealCheckedOut(data.orderId)){
                        if (data.autoRenewing){
                            mCheckout.isPurchasedOneYears = true;
                        }else{
                            mCheckout.isPurchasedOneYears = false;
                        }
                    }else{
                        mCheckout.isPurchasedOneYears = false;
                    }
                }
                Utils.setCheckoutItems(mCheckout);
                break;
            default:
                Utils.setCheckoutItems(new CheckoutItems());
                break;
        }
    }
}

File: app/src/main/java/co/tpcreative/supersafe/common/controller/PremiumManager.kt
package co.tpcreative.supersafe.common.controller
import co.tpcreative.supersafe.R
import co.tpcreative.supersafe.common.extension.getCheckoutItems
import co.tpcreative.supersafe.common.extension.putCheckoutItems
import co.tpcreative.supersafe.common.services.SuperSafeApplication
import co.tpcreative.supersafe.common.util.Utils
import co.tpcreative.supersafe.model.CheckoutItems
import co.tpcreative.supersafe.model.EnumPurchase
import co.tpcreative.supersafe.model.EnumStatus
import com.anjlab.android.iab.v3.BillingProcessor
import com.anjlab.android.iab.v3.PurchaseData
import com.anjlab.android.iab.v3.TransactionDetails
import com.google.gson.Gson

class PremiumManager : BillingProcessor.IBillingHandler {
    private val TAG = PremiumManager::class.java.simpleName

    /*New version*/
    private var bp: BillingProcessor? = null
    fun onStartInAppPurchase() {
        bp = BillingProcessor(SuperSafeApplication.getInstance(), Utils.GOOGLE_CONSOLE_KEY, this)
        bp?.initialize()
    }

    fun onStop() {
        bp?.release()
    }

    /*New version*/
    override fun onProductPurchased(productId: String, details: TransactionDetails?) {}
    override fun onPurchaseHistoryRestored() {
        Utils.Log(TAG,"Restored..........")
    }
    override fun onBillingError(errorCode: Int, error: Throwable?) {}
    override fun onBillingInitialized() {
        /*Checking life time in-app*/
        bp?.let {bpResult ->
            val detailTimeLife: TransactionDetails? = bpResult.getPurchaseTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.life_time))
            val detailSixMonths: TransactionDetails? = bpResult.getSubscriptionTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.six_months))
            val detailsOneYear: TransactionDetails? = bpResult.getSubscriptionTransactionDetails(SuperSafeApplication.getInstance().getString(R.string.one_years))
            if (detailTimeLife != null) {
                val mPurchaseData: PurchaseData = detailTimeLife.purchaseInfo.purchaseData
                onCheckout(mPurchaseData, EnumPurchase.LIFETIME)
            }else if (detailSixMonths!=null) {
                val mPurchaseData: PurchaseData = detailSixMonths.purchaseInfo.purchaseData
                onCheckout(mPurchaseData, EnumPurchase.SIX_MONTHS)
            } else if (detailsOneYear!=null) {
                val mPurchaseData: PurchaseData = detailsOneYear.purchaseInfo.purchaseData
                onCheckout(mPurchaseData, EnumPurchase.ONE_YEAR)
            } else {
                onCheckout(PurchaseData(), EnumPurchase.NONE)
            }
        }
        bp?.loadOwnedPurchasesFromGoogle()
    }

    private fun onCheckout(data: PurchaseData, purchase: EnumPurchase) {
        var mCheckout: CheckoutItems? = Utils.getCheckoutItems()
        Utils.Log(TAG, "Call checkout....")
        Utils.Log(TAG, Gson().toJson(data))
        when (purchase) {
            EnumPurchase.LIFETIME -> {
                if (mCheckout != null) {
                    mCheckout.isPurchasedLifeTime = Utils.isRealCheckedOut(data.orderId)
                } else {
                    mCheckout = CheckoutItems()
                    mCheckout.isPurchasedLifeTime = Utils.isRealCheckedOut(data.orderId)
                }
                Utils.putCheckoutItems(mCheckout)
            }
            EnumPurchase.SIX_MONTHS -> {
                if (mCheckout != null) {
                    if (Utils.isRealCheckedOut(data.orderId)) {
                        mCheckout.isPurchasedSixMonths = data.autoRenewing
                        Utils.Log(TAG,"Checking...6")
                        /*When subscription request to recharge*/
                        if (!mCheckout.isPurchasedSixMonths && Utils.checkingServicesToStopPremiumFeatures()) {
                            Utils.onPushEventBus(EnumStatus.EXPIRED_SUBSCRIPTIONS)
                        }
                    } else {
                        mCheckout.isPurchasedSixMonths = false
                    }
                } else {
                    mCheckout = CheckoutItems()
                    if (Utils.isRealCheckedOut(data.orderId)) {
                        mCheckout.isPurchasedSixMonths = data.autoRenewing
                        /*When subscription request to recharge*/
                        if (!mCheckout.isPurchasedSixMonths && Utils.checkingServicesToStopPremiumFeatures()) {
                            Utils.onPushEventBus(EnumStatus.EXPIRED_SUBSCRIPTIONS)
                        }
                    } else {
                        mCheckout.isPurchasedSixMonths = false
                    }
                }
                Utils.putCheckoutItems(mCheckout)
                Utils.Log(TAG,"Preparing save ${Gson().toJson(mCheckout)}")
            }
            EnumPurchase.ONE_YEAR -> {
                Utils.Log(TAG,"one year...")
                if (mCheckout != null) {
                    if (Utils.isRealCheckedOut(data.orderId)) {
                        mCheckout.isPurchasedOneYears = data.autoRenewing
                        Utils.Log(TAG,"Checking...1y")
                        /*When subscription request to recharge*/
                        if (!mCheckout.isPurchasedOneYears && Utils.checkingServicesToStopPremiumFeatures()) {
                            Utils.onPushEventBus(EnumStatus.EXPIRED_SUBSCRIPTIONS)
                        }
                    } else {
                        mCheckout.isPurchasedOneYears = false
                    }
                } else {
                    mCheckout = CheckoutItems()
                    if (Utils.isRealCheckedOut(data.orderId)) {
                        mCheckout.isPurchasedOneYears = data.autoRenewing
                        /*When subscription request to recharge*/
                        if (!mCheckout.isPurchasedOneYears && Utils.checkingServicesToStopPremiumFeatures()) {
                            Utils.onPushEventBus(EnumStatus.EXPIRED_SUBSCRIPTIONS)
                        }
                    } else {
                        mCheckout.isPurchasedOneYears = false
                    }
                }
                Utils.Log(TAG,"Preparing save ${Gson().toJson(mCheckout)}")
                Utils.putCheckoutItems(mCheckout)
            }
            else -> {
                /*When the user cancel and refund*/
                if (Utils.checkingServicesToStopPremiumFeatures()){
                    Utils.onPushEventBus(EnumStatus.EXPIRED_SUBSCRIPTIONS)
                }
                Utils.checkingExistingSaver()
                Utils.putCheckoutItems(CheckoutItems())
            }
        }
    }


    companion object {
        private var instance: PremiumManager? = null
        fun getInstance(): PremiumManager {
            if (instance == null) {
                instance = PremiumManager()
            }
            return instance!!
        }
    }
}